### PR TITLE
Implement unix time and rfc3339 parsing

### PR DESCRIPTION
**Pull request by colega**
This implements `time` functions `toUnixTimestamp` that calculates a unix timestamp of a date/time, and `parseRFC3339` that does some limited parsing of RFC3339 timestamps.

### CODE REVIEW

File: date.libsonnet [hunk 1]
```diff
@@ -55,4 +55,131 @@ local d = import 'doc-util/main.libsonnet';
...
+    // We don't support timezones, so string should end with 'Z' or 'z'.
+    assert std.endsWith(input, 'Z') || std.endsWith(input, 'z') : 'the provided RFC3339 "%s" should end with "Z" or "z". This implementation does not currently support timezones' % input;
+
+    // RFC3339 can separate date and time using 'T', 't' or ' '.
+    // Find out which one it is and use it.
+    local sep =
+      if stringContains(input, 'T') then 'T'
+      else if stringContains(input, 't') then 't'
+      else if stringContains(input, ' ') then ' '
+      else error 'the provided RFC3339 input "%s" should contain either a "T", or a "t" or space " " as a separator for date and time parts' % input;
```
Comment (colega) [inline date.libsonnet:173]: @Duologic as you asked in https://github.com/grafana/jsonnet-libs/pull/1095, I added support for lowercase t/z and also for space separator.

File: date.libsonnet [hunk 2]
```diff
@@ -55,4 +55,131 @@ local d = import 'doc-util/main.libsonnet';
...
+    sumYearsSeconds(year) + sumMonthsSeconds(year, month) + sumDaysSeconds(day) + hour * 3600 + minute * 60 + second,
+
+  // isNumeric checks that the input is a non-empty string containing only digit characters.
+  local isNumeric(input) =
+    assert std.type(input) == 'string' : 'isNumeric() only operates on string inputs, got %s' % std.type(input);
+    std.foldl(
+      function(acc, char) acc && std.codepoint('0') <= std.codepoint(char) && std.codepoint(char) <= std.codepoint('9'),
+      std.stringChars(input),
+      std.length(input) > 0,
+    ),
```
Comment (colega) [inline date.libsonnet:112]: @jdbaldry I borrowed the implementation from your comment here :)

File: date.libsonnet [hunk 3]
```diff
@@ -55,4 +55,131 @@ local d = import 'doc-util/main.libsonnet';
...
+      d.arg('year', d.T.number),
+      d.arg('month', d.T.number),
+      d.arg('day', d.T.number),
+      d.arg('hour', d.T.number),
+      d.arg('minute', d.T.number),
+      d.arg('second', d.T.number),
+    ],
+  ),
+  toUnixTimestamp(year, month, day, hour, minute, second)::
+    sumYearsSeconds(year) + sumMonthsSeconds(year, month) + sumDaysSeconds(day) + hour * 3600 + minute * 60 + second,
```
Comment (colega) [inline date.libsonnet:103]: This was already in github.com/grafana/jsonnet-libs/date, now I'm bringing here as it's useful once we've parsed the RFC3339.

File: docs/date.md [hunk 4]
```diff
@@ -42,4 +44,23 @@ for common years, and 1-366 for leap years.
...
\ No newline at end of file
+`isLeapYear` returns true if the given year is a leap year.
+
+### fn parseRFC3339
+
+```ts
+parseRFC3339(input)
+```
+
+`parseRFC3339` parses an RFC3339-formatted date & time string into an object containing the 'year', 'month', 'day', 'hour', 'minute' and 'second fields.
```
Comment (malcolmholmes) [inline docs/date.md:55]: Would be useful to have an example of an RFC3339 date/time for those of us that know the format but not the RFC number.

**Review by malcolmholmes — APPROVED** [review]
Looks great to me, very thorough.

File: docs/date.md (hunk 4 above)
Comment (colega) [reply docs/date.md:55]: Sure, followed up in https://github.com/jsonnet-libs/xtd/pull/24

**Pull request merged**